File: API/Domain/Models/Aircraft.cs
using System.ComponentModel.DataAnnotations;
using API.Domain.Models.Enums;

namespace API.Domain.Models
{
  public class Aircraft
  {
    public int Id { get; set; }
    // FK userid
    public string IcaoId { get; set; }
    public string Manufacturer { get; set; }
    public string Model { get; set; }
    public string Variant { get; set; }
    public string Registration { get; set; }
    public EAircraftType AircraftType { get; set; }
    public EEngineType EngineType { get; set; }
    public short EngineCount { get; set; }
    public EWeightCategory WeightCategory { get; set; }
    public EIcaoWakeCategory IcaoWakeCategory { get; set; }
    public EFuelType FuelType { get; set; }
    public int MaxTakeoffWeight { get; set; }
    public int CruiseSpeed { get; set; }
    public decimal FuelCapacity { get; set; }
    // TODO: Convert Fuel Weight <-> Volume
    public decimal MaxRange { get; set; }
    public int ServiceCeiling { get; set; }
    // public int Votes { get; set; }
  }
}

File: Client/web-ui/src/types/Aircraft/Aircraft.ts
/* TODO REMOVE: https://fullstackopen.com/en/part9/typing_the_express_app#utility-types */

import { EAircraftType, EEngineType, EFuelType, EIcaoWakeCategory, EWeightCategory } from "../enums/AircraftEnums";

export interface Aircraft {
  id: string;
  icaoId: string;
  manufacturer: string;
  model: string;
  variant?: string;
  registration?: string;
  aircraftType: EAircraftType;
  engineType: EEngineType;
  engineCount: number;
  weightCategory: EWeightCategory;
  icaoWakeCategory: EIcaoWakeCategory;
  fuelType: EFuelType;
  maxTakeoffWeight: number;
  cruiseSpeed: number;
  fuelCapacity: number;
  maxRange: number;
  serviceCeiling: number;
}

export interface AircraftData extends Aircraft{
  savesCount: number;
  authorUsername?: string; // TODO: Remove optional
}

export interface AircraftState extends AircraftData {
  loadedFuel: number;
}

// Define special omit for unions
type UnionOmit<T, K extends string | number | symbol> = T extends unknown ? Omit<T, K> : never;

export type NewAircraft = UnionOmit<Aircraft, 'id'>;

// TODO: paginated AircraftData
